### PR TITLE
TIP-1229: Add SQL query for project model indexation fetch

### PR DESCRIPTION
**Pull request by pierallard**
Things left:
- all complete
- all incomplete
- attributes of ancestor
- attributes for this level

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+        $query = <<<SQL
+WITH
+    product_model AS (
+        SELECT
+            product_model.id,
+            product_model.code,
+            product_model.created,
+            root_product_model.code AS parent_code,
+            GREATEST(product_model.updated, COALESCE(root_product_model.raw_values, 0)) as updated,
+            JSON_MERGE(COALESCE(root_product_model.raw_values, '{}'), COALESCE(product_model.raw_values, '{}')) AS raw_values,
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: JSON_MERGE_PRESERVE

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        SELECT
+            product_model.id AS product_model_id,
+            JSON_ARRAYAGG(category.code) AS category_codes
+        FROM
+            product_model
+            INNER JOIN pim_catalog_category_product_model category_product_model ON category_product_model.product_model_id = product_model.parent_id
+            INNER JOIN pim_catalog_category category ON category.id = category_product_model.category_id
+        GROUP BY product_model.id
+    ),
+    product_model_family AS (
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: `product_model_family_label` (you will need `family_attributes` later on for the ancestor attributes

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        SELECT 
+            product_model.id AS product_model_id,
+            JSON_ARRAYAGG(JSON_OBJECT(family_translation.locale, family_translation.label)) AS labels
+        FROM 
+            product_model
+            JOIN pim_catalog_family family ON family.code = product_model.family_code
+            JOIN pim_catalog_family_translation family_translation ON family_translation.foreign_key = family.id
+            JOIN pim_catalog_locale locale ON locale.code = family_translation.locale WHERE locale.is_activated = true
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: I would do as I did with a CROSS JOIN to ensure to generate all labels, I don't know if it's useful but it is at least consistent with what we had

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        GROUP BY product_model.id
+    )
+    SELECT
+        product_model.*,
+        COALESCE(product_model_categories.category_codes, JSON_ARRAY()) as category_codes,
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]:
```

        JSON_MERGE_PRESERVE(
            COALESCE(product_model_categories.category_codes, JSON_ARRAY()),
            COALESCE(root_product_model_categories.category_codes, JSON_ARRAY())
        ) AS category_codes,
```

Could you modify the test to cover it also please?

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        $query = <<<SQL
+WITH
+    product_model AS (
+        SELECT
+            product_model.id,
+            product_model.code,
+            product_model.created,
+            root_product_model.code AS parent_code,
+            GREATEST(product_model.updated, COALESCE(root_product_model.raw_values, 0)) as updated,
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: I think a test is important to check the it returns the maximum date, as the business rule is important

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+                'family_code' => Type::getType(Type::STRING)->convertToPHPValue($row['family_code'], $platform),
+                'family_labels' => json_decode($row['family_labels'], true),
+                'family_variant_code' => Type::getType(Type::STRING)->convertToPHPValue($row['family_variant_code'], $platform),
+                'category_codes' => json_decode($row['category_codes']),
+                'ancestor_category_codes' => json_decode($row['ancestor_category_codes']),
+                'parent_code' => $row['parent_code'],
+                'values' => $values,
+                'ancestor_ids' => null !== $row['parent_id'] ? ['product_model_' . $row['parent_id']] : [],
+                'ancestor_codes' => null !== $row['parent_code'] ? [$row['parent_code']] : [],
+                'ancestor_labels' => isset($values[$row['attribute_as_label_code']]) ? $values[$row['attribute_as_label_code']] : [],
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: bug if `$row['attribute_as_label_code']` is null: attribute as label is mandatory but a product without family will have a null value

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+            ['productModelCodes' => Connection::PARAM_STR_ARRAY]
+        );
+
+        $platform = $this->connection->getDatabasePlatform();
+        $results = [];
+        foreach ($rows as $row) {
+            $values = json_decode($row['raw_values'], true);
+
+            $results[$row['code']] = [
+                'id' => Type::getType(Type::INTEGER)->convertToPHPValue($row['id'], $platform),
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: oh you know (int) is enough :smile:

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+                'code' => Type::getType(Type::STRING)->convertToPHPValue($row['code'], $platform),
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: PDO is using string natively, I think it's only useful for date actually

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+                'created' => Type::getType(Type::DATETIME_IMMUTABLE)->convertToPHPValue($row['created'], $platform),
+                'updated' => Type::getType(Type::DATETIME_IMMUTABLE)->convertToPHPValue($row['updated'], $platform),
+                'family_code' => Type::getType(Type::STRING)->convertToPHPValue($row['family_code'], $platform),
+                'family_labels' => json_decode($row['family_labels'], true),
+                'family_variant_code' => Type::getType(Type::STRING)->convertToPHPValue($row['family_variant_code'], $platform),
+                'category_codes' => json_decode($row['category_codes']),
+                'ancestor_category_codes' => json_decode($row['ancestor_category_codes']),
+                'parent_code' => $row['parent_code'],
+                'values' => $values,
+                'ancestor_ids' => null !== $row['parent_id'] ? ['product_model_' . $row['parent_id']] : [],
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: this logic should done in the object in my opinion (but I realized I have a bug on it): it should accept only a single int and construct the array internally.

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+                'ancestor_codes' => null !== $row['parent_code'] ? [$row['parent_code']] : [],
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: same

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: or `JSON_MERGE_PRESERVE`, it does not matter for simple array though

[the remark this answers is not in the document]
Comment (nmarniesse) [reply src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: yes it's JSON_MERGE_PRESERVE in this case, JSON_MERGE_PATCH will not given the expected result here

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php (hunk 6 above)
Comment (pierallard) [reply src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: Impossible ; seen IRL

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Model/ElasticsearchProductModelProjection.php [hunk 11]
```diff
@@ -133,9 +123,9 @@ public function toArray(): array
             'all_complete' => $this->allComplete,
             'all_incomplete' => $this->allIncomplete,
             'ancestors' => [
-                'ids' => $this->ancestorIds,
-                'codes' => $this->ancestorCodes,
-                'labels' => $this->ancestorLabels,
+                'ids' => null !== $this->parentId ? ['product_model_' . (string) $this->parentId] : [],
+                'codes' => null !== $this->parentCode ? [$this->parentCode] : [],
+                'labels' => null !== $this->parentId ? $this->labels : [],
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Model/ElasticsearchProductModelProjection.php]: We don't care the parent labels?

**Comment by ahocquard** [comment]
Missing test on updated date btw

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+        $query = <<<SQL
+WITH
+    product_model AS (
+        SELECT
+            product_model.id,
+            product_model.code,
+            product_model.created,
+            root_product_model.code AS parent_code,
+            GREATEST(product_model.updated, COALESCE(root_product_model.updated, 0)) as updated,
+            JSON_MERGE_PRESERVE(COALESCE(root_product_model.raw_values, '{}'), COALESCE(product_model.raw_values, '{}')) AS raw_values,
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: Sorry, think `JSON_MERGE_PATCH` is better finally for objects.
Give a try:

```sql
SELECT JSON_MERGE_PRESERVE('{"a": "b"}', '{"a": "c"}');
-- {"a": ["b", "c"]} 
SELECT JSON_MERGE_PATCH('{"a": "b"}', '{"a": "c"}');
-- {"a": "c"}
```

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+            LEFT JOIN pim_catalog_family_translation family_translation ON family_translation.foreign_key = family.family_id AND family_translation.locale = locale.code
+        WHERE locale.is_activated = true
+        GROUP BY family.family_id
+    )
+    SELECT
+        product_model.*,
+        JSON_MERGE_PRESERVE(
+            COALESCE(product_model_categories.category_codes, JSON_ARRAY()),
+            COALESCE(root_product_model_categories.category_codes, JSON_ARRAY())
+        ) AS category_codes,
```
Comment (nmarniesse) [inline src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: For arrays, `JSON_MERGE_PRESERVE` is good

File: src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php (hunk 12 above)
Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Bundle/Product/Query/Sql/GetValuesAndPropertiesFromProductModelCodes.php]: There is never the same attribute code in the raw values of root product model and sub product model, so it's the same. If it happens, nothing is better: bad type vs loss of data.
At least bad type will probably raise an exception.

File: src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Model/ElasticsearchProductModelProjection.php (hunk 11 above)
Comment (pierallard) [reply src/Akeneo/Pim/Enrichment/Bundle/Elasticsearch/Model/ElasticsearchProductModelProjection.php]: "parent labels" is exactly the same as "product model label" if it has a parent.

**Pull request merged**